runtime/dynlink_nat.c

```diff
@@ -71,29 +71,42 @@ CAMLprim value caml_natdynlink_open(value filename, value global)
   void *sym;
   void *dlhandle;
   char_os *p;
-
-  /* TODO: dlclose in case of error... */
+  value exn;
 
   p = caml_stat_strdup_to_os(String_val(filename));
-  caml_enter_blocking_section();
+  exn = caml_enter_blocking_section_exn();
+  if (Is_exception_result(exn)) goto cleanup1;
```

Comment:
```suggestion
  if (Is_exception_result(exn)) {
    caml_stat_free(p);
    caml_raise(Extract_exception(exn));
  }
```

Avoiding `goto` when possible is always good. In this case, I don't think the code becomes more complex.